File: GCons/DP1.m
classdef DP1
    %DP1 constraint using r-p formulation. The DP1 constraint exists
    %between a vector aiBar on body i, and a vector ajBar, on body j. DP1
    %constraint specifies that the dot product of the two vectors aiBar
    %and ajBar assume a certian value, F(t), which is frequently a constant
    
    properties
        bodyi; %instance of a body object, i 
        bodyj; %instance of a body object, j
        
        %these point definitions are taken from S10 lecture 9.26
        Qi; %index of a [3x1] marker contained within bodyi, representing head of aiBar
        Pi; %index of a [3x1] marker contained within bodyi, representing tail of aiBar
        Qj; %index of a [3x1] marker contained within bodyj, representing head of ajBar
        Pj; %index of a [3x1] marker contained within bodyj, representing tail of ajBar
        f; %prescribed value that the dot product should assume
        fdot; %derivative of above
        fddot;%derivative of above 
        t;  
    end
    
    %calculated values 
    properties(Dependent)
        aiBar; % vector in local reference frame of body i
        ajBar; % vector in local reference frame of body j
        phi; %jacobian
        nu;  % dPhi_dp1/ dt
        gamma; %RHS of acceleration equation
        phi_ri;%derivative of phi wrt ri
        phi_rj;%derivative of phi wrt rj
        phi_pi;%derivative of phi wrt ri
        phi_pj;%derivative of phi wrt ri
        
        
        
    end
    
    methods
        %constructor
        function dp1 = DP1(bodyi,bodyj,Qi,Pi,Qj,Pj,f,fdot,fddot,t)
            dp1.bodyi = bodyi;
            dp1.bodyj = bodyj;
            dp1.Qi = Qi;
            dp1.Pi = Pi;
            dp1.Qj = Qj;
            dp1.Pj = Pj;
            dp1.f = f;
            dp1.fdot = fdot;
            dp1.fddot = fddot;
            dp1.t = t;
            
        end
        %getters for dependent properties
        
        function aiBar = get.aiBar(dp1) %9.26 slide 10
            %ai vector in the local reference frame
            aiBar = dp1.bodyi.markers{dp1.Qi}-dp1.bodyi.markers{dp1.Pi};
        end
        
        function ajBar = get.ajBar(dp1) %9.26 slide 10
            ajBar = dp1.bodyj.markers{dp1.Qj}-dp1.bodyi.markers{dp1.Pj};
        end
        
        function phi = get.phi(dp1) %9.26 slide 11
            %phi = aiBar'* Ai'* Aj * ajBar
            phi = (dp1.aiBar' * dp1.bodyi.A')*(dp1.bodyj.A * dp1.ajBar) - dp1.f;
        end
        
        function nu = get.nu(dp1) %9.26 slide 12
            nu = dp1.fdot;
        end
        
        function gamma = get.gamma(dp1)% 10.7 slide 8
            %gamma [1x1] is the RHS of the acceleration equation
            ai = dp1.bodyi.A*dp1.aiBar; % a = A*abar
            bpdotj = Matop.B(dp1.bodyj.pdot, dp1.ajBar);
            pdotj = dp1.bodyj.pdot;
            aj = dp1.bodyj.A*dp1.ajBar;
            bpdoti = Matop.B(dp1.bodyi.pdot, dp1.aiBar);
            pdoti = dp1.bodyi.pdot;
            adoti = Matop.B(dp1.bodyi.p, dp1.aiBar)*dp1.bodyi.pdot; %aidot = B(p,abar)*pdot
            adotj = Matop.B(dp1.bodyj.p, dp1.ajBar)*dp1.bodyj.pdot;
            
            
            gamma = -ai'*bpdotj*pdotj - aj'*bpdoti*pdoti - 2*adoti'*adotj + d1.fddot;
            
        end
       
        function phi_ri = get.phi_ri(dp1) % s13 lecture 9.28.2016
            %phi_ri [1x3] is the derivative of phi with respect to the 
            %positional coordinates of bodyi. If I is the ground, then
            %phi_ri is [0x0]
            if dp1.bodyi.isground
                phi_ri = [];
            else
                phi_ri = [0 0 0]
            end
        end
        
        function phi_rj = get.phi_rj(dp1) % s13 lecture 9.28.2016
            %phi_ri [1x3] is the derivative of phi with respect to the 
            %positional coordinates of bodyi. If I is the ground, then
            %phi_rj is [0x0]
            if dp1.bodyj.isground
                phi_rj = [];
            else
                phi_rj = [0 0 0];
            end
        end
       
        
        function phi_pi = get.phi_pi(dp1) % s13 lecture 9.28.2016
            %phi_pi [1x4] is the derivative of phi with respect to the 
            %orientational coordinates of bodyi. If bodyi is the ground
            %then phi_pi is [0x0]
            if dp1.bodyi.isground
                phi_pi = []
            else
                phi_pi = (dp1.bodyj.A*dp1.ajBar)'*MatOp.B(dp1.bodyi.p, dp1.aiBar);
            end
        end
        
        function phi_pj = get.phi_pj(dp1) % s13 lecture 9.28.2016
            %phi_pi [1x4] is the derivative of phi with respect to the 
            %orientational coordinates of bodyi. If bodyi is the ground
            %then phi_pi is [0x0]
            if dp1.bodyj.isground
                phi_pj = [];
            else
                phi_pj =(dp1.bodyi.A*dp1.aiBar)'*MatOp.B(dp1.bodyj.p, dp1.ajBar);
            end
            
            end
        end
           
            
    
end

File: julia/GCons/dp1.jl
#GCons are Geometic Constraints
include("..\\body.jl") #this needs work !!
type dp1
  """
  The dp1 or dot product 1 constraint reflects the fact that the dot product of
  between a vector on body i and a second vector on body j assume a specified value
  d1 is one of 4 basic GCons that each remove one DOF
  """
  sim::Sim      #reference to the simulation data structures
  bodyi::Body   #bodyi
  bodyj::Body   #bodyj
  Pi::Int       #index of point p on body i (tail)
  Qi::Int       #index of point q on body i (head)
  Pj::Int       #index of point P on body j (tail)
  Qj::Int       #index of point Q on body j (head)
  rDOF::Int     #number of degrees of freedom removed by constraint
  f             #lambda function describing how constraint changes with time
  fdot          #lambda function describing how constraint changes with time
  fddot         #lambda function describing how constraint changes with time



  #constructor function
  function dp1(sim::Sim,bi::Body,bj::Body,Qi::Int,Qj::Int,Pi=1,Pj=1,f = t->0 , fdot = t->0, fddot = t->0)
    rDOF = 1; #dp1 removes one DOF
    new(sim,bi,bj,Pi,Qi,Pj,Qj,rDOF,f,fdot,fddot) #make a new dp1
  end
end

#----------------begin functions associated with dp1----------------------------
#pseudo - getter methods.
aBari(con::dp1) = pt(con.bodyi,con.Qi) - pt(con.bodyi,Pi) #[3x1]
aBarj(con::dp1) = pt(con.bodyi,con.Qj) - pt(con.bodyi,Pj) #[3x1]


function ϕ(con::dp1)   #9.26.2016 - slide 11
  """
  constraint equation ϕ
  output: [1 x 1] evaluation of constraint equation value
  """
  phi = aBari(con)'*A(con.bodyi)'*A(con.bodyj)*aBarj(con) - con.f(con.sim.t)
end

function ν(con::dp1)  #9.26.2016 - slide 12
  """
  RHS of vel equation
  output: [1 x 1] evaluation ν
  """
  nu = con.fdot(con.sim.t)
end

function 	γ(con::dp1)  #10.7.2016 - slide 8
"""
RHS of accel equation
output: [1 x 1] evaluation ν
"""
aiBar = aBari(con) ; ajBar = aBarj(con)
ai = A(con.bodyi)*aiBar ; aj = A(cons.bodyj)*ajBar
aidot =B(p(con.bodyi) ,aiBar)*pdot(con.bodyi)
ajdot =B(p(con.bodyj) ,ajBar)*pdot(con.bodyj)
pdoti = pdot(con.bodyi) ; pdotj = pdot(con.bodyj)

gamma = -ai'B(pdotj,ajBar)*pdotj - aj'B(pdoti,aiBar)*pdoti - 2*aidot'ajdot + fddot(con.sim.t)
end

function ϕ_r(con::dp1)  #9.28.2016 slide 13
  """
  partial derivative of ϕ WRT position position GC's of both bodyi and bodyj
  output: ([1,3],[1,3]) vectors of
  """
  phi_ri = zero(1,3) ; phi_rj = zeros(1,3)
  return phi_ri , ph_rj
end

function ϕ_p(con::dp1)  # #9.28.2016 slide 13
"""
partial derivative of ϕ WRT position orientation GC's of both bodyi and bodyj
output:([1,4],[1,4])
"""
phi_pi = aBarj(con)'*A(con.bodyj)'*B(p(cons.bodyi),aBari(con))
phi_pj = aBari(con)'*A(con.bodyi)'*B(p(cons.bodyj),aBarj(con))
return phi_pi , phi_pj
end
